sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala

```diff
@@ -521,38 +522,81 @@ class Analyzer(
    */
   object ResolveSortReferences extends Rule[LogicalPlan] {
     def apply(plan: LogicalPlan): LogicalPlan = plan resolveOperators {
-      case s @ Sort(ordering, global, p @ Project(projectList, child))
-          if !s.resolved && p.resolved =>
-        val (newOrdering, missing) = resolveAndFindMissing(ordering, p, child)
+      case s @ Sort(_, _, child) if !s.resolved && child.resolved =>
+        val (newOrdering, missingAttrs, missingPlans) =
+          collectResolvedMissingAttrs(s.order, child, Seq.empty[LogicalPlan])
 
-        // If this rule was not a no-op, return the transformed plan, otherwise return the original.
-        if (missing.nonEmpty) {
-          // Add missing attributes and then project them away after the sort.
-          Project(p.output,
-            Sort(newOrdering, global,
-              Project(projectList ++ missing, child)))
-        } else {
-          logDebug(s"Failed to find $missing in ${p.output.mkString(", ")}")
+        if (missingAttrs.isEmpty) {
+          logDebug(s"Failed to find $missingAttrs in ${child.output.mkString(", ")}")
           s // Nothing we can do here. Return original plan.
         }
+        else {
+          val missingPlanSet = missingPlans.toSet
```

Comment:
Thank you! I will change the implementation. @davies 

Originally, I thought we were not allowed to use a `var` flag since all the rules in optimizer and analyzer do not use it. Now, if we can do it like that, it is much simpler to do the tree traversal. : )